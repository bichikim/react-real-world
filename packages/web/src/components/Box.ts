import styled from '@emotion/styled'
import {FC, createElement as h, useCallback} from 'react'
import {PureObject} from 'src/types'
import {useDebounce, withFastMemo} from 'src/hooks'
import {
  color,
  ColorProps,
  flexbox, FlexboxProps,
  grid,
  GridProps,
  layout,
  LayoutProps,
  position,
  PositionProps,
  ResponsiveValue,
  shadow,
  ShadowProps,
  space,
  SpaceProps,
  style,
  typography,
  TypographyProps,
} from 'styled-system'
import {createShouldForwardProp, props} from '@styled-system/should-forward-prop'

interface TransformProps {
  transform?: ResponsiveValue<string>
}

const transform = style({
  cssProperty: 'transform',
  prop: 'transform',
})

const StyledBox: FC<PureObject> = styled('div', {shouldForwardProp: createShouldForwardProp([...props, 'transform'])})(
  color, space, layout, position, shadow, flexbox, grid, typography, transform,
)

export interface ASProps {
  as?: string
}

export interface HtmlProps {
  disabled?: boolean
  onClick?: () => any
  type?: string
}

export interface ImgProps {
  src?: string
}

export interface HtmlInputProps<S = any> {
  checked?: boolean
  onChange?: (value: S) => any
  value?: S
}

export interface BoxProps<S = any> extends ColorProps, SpaceProps, LayoutProps, PositionProps, ShadowProps,
  FlexboxProps, GridProps, TypographyProps, HtmlInputProps<S>, HtmlProps, ASProps, ImgProps, TransformProps {
}

const ASYNC_WAIT = 45

const useAsyncEvent = (onChange, getter?: (value: any) => any, wait:number = ASYNC_WAIT) => {
  return useDebounce(onChange, wait, {getter})
}

const useChange = (onChange, type) => {
  return useCallback((event) => {
    switch (type) {
      case 'checkbox':
        return onChange && onChange(event?.target?.checked)
      case 'radio':
        return onChange && onChange(event?.target?.checked)
      default:
        return onChange && onChange(event?.target?.value)
    }
  }, [onChange, type])
}

export const Box: FC<BoxProps> = withFastMemo((props) => {
  const {children, value = '', type, ...rest} = props

  const onClick = useAsyncEvent(props.onClick)
  const onChange = useChange(props.onChange, type)

  return (
    h(StyledBox, {...rest, onChange, onClick, type, value}, children)
  )
})
