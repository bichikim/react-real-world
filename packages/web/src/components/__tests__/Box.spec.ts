import {act, fireEvent, render} from '@testing-library/react'
import {createElement as h} from 'react'
import {SinonFakeTimers, useFakeTimers} from 'sinon'
import {Box} from '../Box'

const TEN = 10
const BG = 'blue'

const setup = () => {
  const onClick = jest.fn(() => {
    return null
  })
  const wrapper = render(h(Box, {
    bg: BG, boxShadow: '0 0 0 1ex black',
    display: 'grid', gridTemplateColumns: '200px, 400px, 200px',
    m: TEN, onClick, p: TEN,
    position: 'absolute', size: TEN,
  }))

  return {
    ...wrapper,
    onClick,
  }
}

describe('Box', function test() {

  let clock: SinonFakeTimers

  beforeEach(() => {
    clock = useFakeTimers()
  })

  afterEach(() => {
    clock.restore()
  })

  it('should render it', function test() {
    const {container} = setup()
    expect(container.firstChild).toMatchSnapshot()
  })

  it('should call onClick asynchronously', function test() {
    const {container, onClick} = setup()

    act(() => {
      fireEvent.click(container.firstChild)
    })

    expect(onClick.mock.calls.length).toBe(0)

    act(() => {
      clock.tick(50)
    })

    expect(onClick.mock.calls.length).toBe(1)
  })
})
