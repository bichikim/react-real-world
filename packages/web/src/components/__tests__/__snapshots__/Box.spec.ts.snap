// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Box should render it 1`] = `
.emotion-0 {
  background-color: blue;
  margin: 10px;
  padding: 10px;
  display: grid;
  width: 10px;
  height: 10px;
  position: absolute;
  box-shadow: 0 0 0 1ex black;
  grid-template-columns: 200px,400px,200px;
}

<div
  class="emotion-0"
  display="grid"
  size="10"
  value=""
/>
`;
